### PR TITLE
refactor(src): split app base into smaller libs

### PR DESCRIPTION
**Pull request by kstdl**
## Purpose
Split massive app_base lib into a smaller libs. 
Move configs into a separate files in a config lib to reduce dependencies.
Merge common and util libs into a one 

## For reviewers
It will be simplier to review it by commits, not like all changes list 

## Deployment Considerations
We need to check the container and binary size because all libs from src is static for now. I compared it to resulting container from develop branc and seems like it become smaller, but i saw a massive cache thing in DRIC: save cache to registry step

### CODE REVIEW

File: src/config/include/config/final_chain.hpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Comment (MatusKysel) [inline src/config/include/config/final_chain.hpp:1]: Uff, please change the name of the file, because now we have 2 files with the same name

File: src/config/include/config/state_api.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
+#pragma once
```
Comment (MatusKysel) [inline src/config/include/config/state_api.hpp:1]: Same issue with the name here

**Review by JakubFornadel — CHANGES_REQUESTED** [review]
So a few notes:

- src/cli/src/config.cpp does not seem the best structure with redundant src. How about this structure:
libraries/
     &nbsp;example_lib/
           src/
           include/example_lib/
programs/
    taraxad/
    
- as we have all headers in include subfolders, we could take advantage of it and create list of headers like this: file(GLOB HEADERS "include/*.hpp")
- I also saw in multiple projects that they use #include <...> for libraries headers even if those libs are part of the project but that is just something to think about :)
- What is app_base INTERFACE good for now when it does not have any source files ? We can link  final binaries against these smaller libs as some tests for example might need only few of them etc... 
- why has every library hardcoded STATIC linking ? 
- does this `target_include_directories(config PUBLIC "include")` work without any dependency on directory from which you trigger cmake ? Somethink like this might be more explicit and clear `${CMAKE_CURRENT_SOURCE_DIR}/include`

File: src/network/src/network.cpp [hunk 3]
```diff
@@ -1,12 +1,12 @@
-#include "network.hpp"
+#include "network/network.hpp"
```
Comment (MatusKysel) [inline src/network/src/network.cpp:1]: `src/network/network.cpp → src/network/src/network.cpp` I do not understand why we need one more time `src` there, why not  `src/network/network.cpp`

Comment (MatusKysel) [reply src/network/src/network.cpp:1]: And also if there is reason for that why files like `src/network/rpc/WSServer.cpp` are not part of `src`

File: src/config/include/config/final_chain.hpp (hunk 1 above)
Comment (kstdl) [reply src/config/include/config/final_chain.hpp:1]: why is it a problem?

File: src/network/src/network.cpp (hunk 3 above)
Comment (kstdl) [reply src/network/src/network.cpp:1]: My further plan is to rename get rid of root src and move all that libs to libraries dir as Jakub suggested above. But I planned to make it in a separate PR right after it to discuss it separately. So every lib will have its own src and there will be no mess in library root dir. You will see CMakeLists.txt, src and include directories in every lib dir

File: src/config/include/config/final_chain.hpp (hunk 1 above)
Comment (MatusKysel) [reply src/config/include/config/final_chain.hpp:1]: I just find it confusing to have 2 files with the name `final_chain.hpp` when one is actually config for another one

Comment (kstdl) [reply src/config/include/config/final_chain.hpp:1]: it is located in another directory and a lib. So you will see `config/final_chain.hpp`. Not sure that `config/final_chain_config.hpp` is better way to do it

Comment (MatusKysel) [reply src/config/include/config/final_chain.hpp:1]: I think better way would be to have config per library and not library that consist of all configs as libraries should be independent

Comment (JakubFornadel) [reply src/config/include/config/final_chain.hpp:1]: I agree with Matus that every libraries should have it's own config as these libraries should (in ideal world) be relatively independent. Why should config for final_chain lib be in different lib ?

Comment (kstdl) [reply src/config/include/config/final_chain.hpp:1]: That things is causing circular dependencies because that configs are needed in other libs and sometimes depends on each other. We still have a lot of it in project, I just started to deal with it in that PR

**Comment by kstdl** [comment]
@JakubFornadel will make reply to you notes in a separate comments that it will be easy to discuss
> * src/cli/src/config.cpp does not seem the best structure with redundant src. How about this structure:
>   libraries/
>       example_lib/
>           src/
>           include/example_lib/
>           CMakeLists.txt
>   programs/
>       taraxad/

I planned to make just that structure(and we discussed it like a week ago or so), but I planned to do it in a separate PR just after that one

**Comment by kstdl** [comment]
> * as we have all headers in include subfolders, we could take advantage of it and create list of headers like this: file(GLOB HEADERS "include/*.hpp")

I think better way to specify it explicitly. But it could be fixed easily. Also we could do it for sources also

**Comment by kstdl** [comment]
> * why has every library hardcoded STATIC linking ?

It couldn't be shared because of circular dependencies. Don't think its a big problem for now. It didn't affect binary size much as I checked

**Comment by MatusKysel** [comment]
> > * why has every library hardcoded STATIC linking ?
> 
> It couldn't be shared because of circular dependencies. Don't think its a big problem for now. It didn't affect binary size much as I checked

actually there is huge difference

before

```
total 169668
drwxrwxr-x  2 matus matus     4096 júl  1 12:56 ./
drwxrwxr-x 10 matus matus     4096 júl  1 12:53 ../
-rwxrwxr-x  1 matus matus  9441688 júl  1 12:55 crypto_test*
-rwxrwxr-x  1 matus matus 11902584 júl  1 12:55 dag_block_test*
-rwxrwxr-x  1 matus matus 11707720 júl  1 12:56 dag_test*
-rwxrwxr-x  1 matus matus  9426480 júl  1 12:56 final_chain_test*
-rwxrwxr-x  1 matus matus 16363720 júl  1 12:55 full_node_test*
-rwxrwxr-x  1 matus matus 13909568 júl  1 12:55 network_test*
-rwxrwxr-x  1 matus matus 10871344 júl  1 12:55 pbft_chain_test*
-rwxrwxr-x  1 matus matus 11393256 júl  1 12:56 pbft_manager_test*
-rwxrwxr-x  1 matus matus 11816776 júl  1 12:55 p2p_test*
-rwxrwxr-x  1 matus matus    18488 júl  1 12:53 reference*
-rwxrwxr-x  1 matus matus  6905912 júl  1 12:56 replay_protection_service_test*
-rwxrwxr-x  1 matus matus 10779032 júl  1 12:55 rpc_taraxa_test*
-rwxrwxr-x  1 matus matus 11124496 júl  1 12:56 state_api_test*
-rwxrwxr-x  1 matus matus 14624784 júl  1 12:55 taraxad*
-rwxrwxr-x  1 matus matus 11925080 júl  1 12:56 transaction_test*
-rwxrwxr-x  1 matus matus 11488512 júl  1 12:55 vote_test*
```

after your changes:
```
total 3483632
drwxrwxr-x  2 matus matus      4096 júl  1 12:39 ./
drwxrwxr-x 10 matus matus      4096 júl  1 12:28 ../
-rwxrwxr-x  1 matus matus 155527368 júl  1 12:37 crypto_test*
-rwxrwxr-x  1 matus matus 289450360 júl  1 12:37 dag_block_test*
-rwxrwxr-x  1 matus matus 165332536 júl  1 12:39 dag_test*
-rwxrwxr-x  1 matus matus 187203000 júl  1 12:39 final_chain_test*
-rwxrwxr-x  1 matus matus 291119368 júl  1 12:38 full_node_test*
-rwxrwxr-x  1 matus matus 289126576 júl  1 12:37 network_test*
-rwxrwxr-x  1 matus matus 286043544 júl  1 12:37 pbft_chain_test*
-rwxrwxr-x  1 matus matus 286854608 júl  1 12:38 pbft_manager_test*
-rwxrwxr-x  1 matus matus 226684664 júl  1 12:37 p2p_test*
-rwxrwxr-x  1 matus matus     18488 júl  1 12:28 reference*
-rwxrwxr-x  1 matus matus 145625312 júl  1 12:38 replay_protection_service_test*
-rwxrwxr-x  1 matus matus 285892712 júl  1 12:36 rpc_taraxa_test*
-rwxrwxr-x  1 matus matus 155842728 júl  1 12:39 state_api_test*
-rwxrwxr-x  1 matus matus 289069888 júl  1 12:36 taraxad*
-rwxrwxr-x  1 matus matus 226430880 júl  1 12:38 transaction_test*
-rwxrwxr-x  1 matus matus 286944728 júl  1 12:36 vote_test*
```

and @JakubFornadel renabled shared libs, so our docker image with tests is not enormous

**Comment by JakubFornadel** [comment]
> > * why has every library hardcoded STATIC linking ?
> 
> It couldn't be shared because of circular dependencies. Don't think its a big problem for now. It didn't affect binary size much as I checked

What is the purpose of sub-library if it has circular dependency on some other library then ? This hardcoded STATIC is pretty big NOGO for me. 

> @JakubFornadel will make reply to you notes in a separate comments that it will be easy to discuss
> 
> > * src/cli/src/config.cpp does not seem the best structure with redundant src. How about this structure:
> >   libraries/
> >       example_lib/
> >           src/
> >           include/example_lib/
> >           CMakeLists.txt
> >   programs/
> >       taraxad/
> 
> I planned to make just that structure(and we discussed it like a week ago or so), but I planned to do it in a separate PR just after that one

File: src/network/src/network.cpp (hunk 3 above)
Comment (JakubFornadel) [reply src/network/src/network.cpp:1]: why not to do it straight away ?

**Comment by kstdl** [comment]
Also just noticed that it affected build time much. Converted to a draft

**Comment by kstdl** [comment]
@MatusKysel @JakubFornadel @gaoqi8649 @mfrankovi you can review new structure
I need to rebase it to latest develop(have much conflict there right now) and remove app_base interface library after

File: libraries/config/include/config/version.hpp [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#pragma once
```
Comment (MatusKysel) [inline libraries/config/include/config/version.hpp:1]: this should not be commited

Comment (kstdl) [reply libraries/config/include/config/version.hpp:1]: already removed it during rebase

**Comment by JakubFornadel** [comment]
> @MatusKysel @JakubFornadel @gaoqi8649 @mfrankovi you can review new structure
> I need to rebase it to latest develop(have much conflict there right now) and remove app_base interface library after

Well first we have to decide if we merge my changes first of yours. Either way it is going to be quite painful to merge it but if you resolve conflicts with develop now -> then I will have to resolve the same conflicts again (just manually) as I have used old develop for copy pasting code into separate handlers....  and because it is in different files now it wont matter that you already solved it...

**Comment by kstdl** [comment]
> Well first we have to decide if we merge my changes first or yours. Either way it is going to be quite painful to merge it but if you resolve conflicts with develop now -> then I will have to resolve the same conflicts again (just manually) as I have used old develop for copy pasting code into separate handlers.... and because it is in different files now it wont matter that you already solved it... Depends how much conflicts is inside tarcap related files ?

I have conflict in other files now, not much related to the network. If your PR is almost ready we can merge it and I will rebase my branch after

File: .gitignore [hunk 5]
```diff
@@ -1,4 +1,5 @@
 .vscode
+compile_commands.json
```
Comment (JakubFornadel) [inline .gitignore:2]: what is this file ?

File: libraries/aleth/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,16 @@
+### Aleth part
```
Comment (JakubFornadel) [inline libraries/aleth/CMakeLists.txt:1]: hmm when we finally have these sub-libraries. Do we even need aleth interface ? We could link separate devcore devcrypto p2p libs when needed, so it should be even more effective...

File: tests/dag_block_test.cpp [hunk 7]
```diff
@@ -194,29 +193,6 @@ TEST_F(DagBlockTest, push_and_pop) {
...
-  EXPECT_TRUE(overlap1[0]);
-  EXPECT_TRUE(overlap1[1]);
-  EXPECT_TRUE(overlap1[2]);
-
-  EXPECT_TRUE(overlap2[0]);
-  EXPECT_FALSE(overlap2[1]);
-  EXPECT_FALSE(overlap2[2]);
-  EXPECT_FALSE(overlap2[3]);
-}
-
```
Comment (MatusKysel) [inline tests/dag_block_test.cpp:219]: What's wrong with this test?

File: libraries/node/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,55 @@
+file(GLOB_RECURSE CONSENSUS_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/consensus/*.hpp)
+file(GLOB_RECURSE CONSENSUS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/consensus/*.cpp)
+
+file(GLOB_RECURSE NETWORK_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/network/*.hpp)
+file(GLOB_RECURSE NETWORK_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/network/*.cpp)
```
Comment (JakubFornadel) [inline libraries/node/CMakeLists.txt:5]: So network and consensus are the last parts that should be "logically" separated into the libs... I guess it was not possible due to some major problems, e.g. circular deps, etc... that is too difficult to fix ?

File: libraries/storage/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+add_library(storage ${SOURCES} ${HEADERS})
+target_include_directories(storage PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
+target_link_libraries(storage PUBLIC CONAN_PKG::rocksdb)
+target_link_libraries(storage PRIVATE 
+    common
+    config
+    logger
+    pbft_block
+    proposal_period_levels_map
```
Comment (MatusKysel) [inline libraries/storage/CMakeLists.txt:17]: does this really needs to be separate lib?

File: libraries/storage/include/storage/storage.hpp [hunk 10]
```diff
@@ -135,9 +131,9 @@ class DbStorage : public std::enable_shared_from_this<DbStorage> {
   const std::string db_dir = "db";
   const std::string state_db_dir = "state_db";
   rocksdb::DB* db_;
-  vector<ColumnFamilyHandle*> handles_;
-  ReadOptions read_options_;
-  WriteOptions write_options_;
+  vector<rocksdb::ColumnFamilyHandle*> handles_;
```
Comment (MatusKysel) [inline libraries/storage/include/storage/storage.hpp:134]: std::vector

File: libraries/storage/include/storage/storage.hpp [hunk 11]
```diff
@@ -412,7 +403,7 @@ class DbStorage : public std::enable_shared_from_this<DbStorage> {
 
   class MultiGetQuery {
     shared_ptr<DbStorage> const db_;
-    vector<ColumnFamilyHandle*> cfs_;
+    vector<rocksdb::ColumnFamilyHandle*> cfs_;
```
Comment (MatusKysel) [inline libraries/storage/include/storage/storage.hpp:406]: std::

File: libraries/types/pbft_block/include/pbft/sync_block.hpp [hunk 12]
```diff
@@ -28,6 +29,9 @@ class SyncBlock {
   std::vector<Transaction> transactions;
   bytes rlp() const;
   void clear();
+  bool hasEnoughValidCertVotes(size_t valid_sortition_players, size_t sortition_threshold, size_t pbft_2t_plus_1) const;
+
+  LOG_OBJECTS_DEFINE
```
Comment (MatusKysel) [inline libraries/types/pbft_block/include/pbft/sync_block.hpp:34]: I do not like loggers inside small class that will be created N times, just necessary overhead. I think better solution is to throw exception and log it later

File: libraries/types/vote/include/vote/vote.hpp [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+  blk_hash_t sha3(bool inc_sig) const { return dev::sha3(rlp(inc_sig)); }
+
+  vote_hash_t vote_hash_;  // hash of this vote
+  blk_hash_t blockhash_;   // Voted PBFT block hash
+  sig_t vote_signature_;
+  VrfPbftSortition vrf_sortition_;
+  mutable public_t cached_voter_;
+  mutable addr_t cached_voter_addr_;
+
+  LOG_OBJECTS_DEFINE
```
Comment (MatusKysel) [inline libraries/types/vote/include/vote/vote.hpp:142]: same what I commented for sync_block and logging

File: libraries/types/proposal_period_levels_map/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,8 @@
+set(HEADERS include/dag/proposal_period_levels_map.hpp)
```
Comment (MatusKysel) [inline libraries/types/proposal_period_levels_map/CMakeLists.txt:1]: as I have already commented I do not think this should be separate lib

File: libraries/storage/src/storage.cpp [hunk 15]
```diff
@@ -40,9 +43,10 @@ DbStorage::DbStorage(fs::path const& path, uint32_t db_snapshot_each_n_pbft_bloc
   options.create_missing_column_families = true;
   options.create_if_missing = true;
   options.compression = rocksdb::CompressionType::kLZ4Compression;
-  vector<ColumnFamilyDescriptor> descriptors;
-  std::transform(Columns::all.begin(), Columns::all.end(), std::back_inserter(descriptors),
-                 [](const Column& col) { return ColumnFamilyDescriptor(col.name(), ColumnFamilyOptions()); });
+  vector<rocksdb::ColumnFamilyDescriptor> descriptors;
```
Comment (MatusKysel) [inline libraries/storage/src/storage.cpp:46]: std::

**Comment by JakubFornadel** [comment]
I am just brainstorming now and trying to figure out how to "sort of separate" network/consensus and node libraries. I get it that it is probably too difficult to separate it as standalone libraries, right ? Isn't it somehow possible to create some sort of interface library that would link against these libraries (which are not buildable as standalone libs) but it could be built altogether ? In such case we could have them inside the separate folders with their own cmakefiles, but libraries/CMakeLists.txt would not include them, it would include only this interface lib that link against them and can be built ?

File: libraries/node/network/rpc/Test.cpp [hunk 16]
```diff
@@ -5,11 +5,11 @@
 #include <libdevcore/CommonJS.h>
```
Comment (MatusKysel) [inline libraries/node/network/rpc/Test.cpp:5]: As I can not comment on previous files (auto generated), did you also change scripts that are creating them? please check  `jsonrpcstub.cmake` and related files. To check if files are correctly generated, I think you need to do `make jsonrpcstub` and then run that binary or something like that

**Comment by kstdl** [comment]
> I am just brainstorming now and trying to figure out how to "sort of separate" network/consensus and node libraries. I get it that it is probably too difficult to separate it as standalone libraries, right ? Isn't it somehow possible to create some sort of interface library that would link against these libraries (which are not buildable as standalone libs) but it could be built altogether ? In such case we could have them inside the separate folders with their own cmakefiles, but libraries/CMakeLists.txt would not include them, it would include only this interface lib that link against them and can be built ?

I searched for it and I have not found any other adequate solution. Tried to do it with OBJECT libraries(kind of you describing), but get an error that libs with cyclic(circular) dependencies could be built only static. Maybe this could be done in some tricky way, but I don't think it's worth it

File: libraries/node/consensus/src/transaction_manager/transaction_manager.cpp [hunk 17]
```diff
@@ -121,16 +116,6 @@ std::pair<bool, std::string> TransactionManager::insertTransaction(const Transac
     }
   }
 
-  // TODO: new pending tx could be omitted here ? Currently we ommit new pending tx only if it is included in dag block
-  // TODO: either create different events, e.g. new pending vs new accepted or remove this whole TODO
-  //  transaction_accepted_.emit(trx.getHash());
```
Comment (MatusKysel) [inline libraries/node/consensus/src/transaction_manager/transaction_manager.cpp:126]: why was this removed?

File: libraries/node/consensus/src/transaction_manager/transaction_manager.cpp [hunk 18]
```diff
@@ -121,16 +116,6 @@ std::pair<bool, std::string> TransactionManager::insertTransaction(const Transac
...
-
-  if (broadcast == true) {
-    if (auto net = network_.lock(); net && conf_.network.network_transaction_interval == 0) {
-      net->onNewTransactions({trx});
```
Comment (MatusKysel) [inline libraries/node/consensus/src/transaction_manager/transaction_manager.cpp:130]: where is this functionality now?

File: .gitignore (hunk 5 above)
Comment (kstdl) [reply .gitignore:2]: File exported by cmake. Needed for some indexing tools(clangd for example). It should be in the workspace to work properly

File: libraries/aleth/CMakeLists.txt (hunk 6 above)
Comment (kstdl) [reply libraries/aleth/CMakeLists.txt:1]: I will check it one more time, but I think we are using all this libs in common. So all other libs also need them all

File: libraries/node/consensus/include/vote_manager/vote_manager.hpp [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+  void verifyVotes(uint64_t pbft_round, size_t sortition_threshold, uint64_t dpos_total_votes_count,
+                   std::function<size_t(addr_t const&)> const& dpos_eligible_vote_count);
+
+  void cleanupVotes(uint64_t pbft_round);
+
+  bool voteValidation(std::shared_ptr<Vote>& vote, size_t valid_sortition_players, size_t sortition_threshold) const;
+
+  bool pbftBlockHasEnoughValidCertVotes(SyncBlock& pbft_block_and_votes, size_t valid_sortition_players,
+                                        size_t sortition_threshold, size_t pbft_2t_plus_1) const;
```
[the remark this answers is not in the document]
Comment (kstdl) [reply libraries/node/consensus/include/vote_manager/vote_manager.hpp:100]: yeah. missed that during rebase

**Comment by JakubFornadel** [comment]
> > I am just brainstorming now and trying to figure out how to "sort of separate" network/consensus and node libraries. I get it that it is probably too difficult to separate it as standalone libraries, right ? Isn't it somehow possible to create some sort of interface library that would link against these libraries (which are not buildable as standalone libs) but it could be built altogether ? In such case we could have them inside the separate folders with their own cmakefiles, but libraries/CMakeLists.txt would not include them, it would include only this interface lib that link against them and can be built ?
> 
> I searched for it and I have not found any other adequate solution. Tried to do it with OBJECT libraries(kind of you describing), but get an error that libs with cyclic(circular) dependencies could be built only static. Maybe this could be done in some tricky way, but I don't think it's worth it

yeah I though so :/ In that case I would personally go with some more name, something like we had before - `app_base` or something else just something generic because if a library is called node I would not expect to find inside network stuff etc... On the other hand if I see app_base it evokes the feeling that it groups together multiple parts... We could have app_base library and internal folder structure would look like this:

```
- libraries/app_base
   - CMakeLists.txt
   - network/
       - include/
       - src/   
   - consensus/
       - include/
       - src/   
    - node/
       - include/
       - src/    
```

with big toto inside CMakeLists.txt that these folders are candidates for separate libs in case someone would be brave enough in the future :D

File: libraries/node/consensus/include/final_chain/data.hpp [hunk 20]
```diff
@@ -36,10 +37,22 @@ struct BlockHeader {
...
   HAS_RLP_FIELDS
 
-  static h256 const& uncles_hash();
-  static Nonce const& nonce();
-  static u256 const& difficulty();
-  static h256 const& mix_hash();
+  static h256 const& uncles_hash() { return EmptyRLPListSHA3(); }
+
+  static Nonce const& nonce() {
+    static const Nonce ret;
```
Comment (MatusKysel) [inline libraries/node/consensus/include/final_chain/data.hpp:43]: correct me if I am wrong but this will created separate static variable for each file where it is included, so I find this absolutely in a conflict why was it coded like that. It needs to go to cpp file

File: libraries/node/CMakeLists.txt (hunk 8 above)
Comment (kstdl) [reply libraries/node/CMakeLists.txt:5]: First part is already discussed below I think.
I decided to make a structure like these to not mix all that files in common include/src directories and highlight that it should be a separate libs. But your proposal on structure is better except that `app_base` name :D

File: libraries/storage/CMakeLists.txt (hunk 9 above)
Comment (kstdl) [reply libraries/storage/CMakeLists.txt:17]: That was a part of the dag. But it is needed for `storage` lib. The closest thing it could be merged with is `dag_block`. So I can do it if you want

File: libraries/node/CMakeLists.txt (hunk 8 above)
Comment (JakubFornadel) [reply libraries/node/CMakeLists.txt:5]: yop feel free to come up with better naming just make sure it is something evoking that it consist of multiple libs....

File: CMakeModules/clang_format.cmake [hunk 21]
```diff
@@ -13,8 +13,10 @@ else ()
 
     # get all project files
     file(GLOB_RECURSE ALL_SOURCE_FILES
-            ${CMAKE_CURRENT_SOURCE_DIR}/src/*.[ch]pp
-            ${CMAKE_CURRENT_SOURCE_DIR}/src/*.h # jsonrpc stubs are generated as .h
+            ${CMAKE_CURRENT_SOURCE_DIR}/libraries/*.[ch]pp
+            ${CMAKE_CURRENT_SOURCE_DIR}/libraries/*.h # jsonrpc stubs are generated as .h
+            ${CMAKE_CURRENT_SOURCE_DIR}/programs/*.[ch]pp
+            ${CMAKE_CURRENT_SOURCE_DIR}/programs/*.h # jsonrpc stubs are generated as .h
```
Comment (JakubFornadel) [inline CMakeModules/clang_format.cmake:19]: no file inside programs will have .h suffix

File: libraries/node/consensus/include/final_chain/data.hpp (hunk 20 above)
Comment (JakubFornadel) [reply libraries/node/consensus/include/final_chain/data.hpp:43]: Do you guys understand what are these 3 static methods even good for because I dont ? :D Seems like it is supposed to return the same default value every time. In such case I would maybe change naming to default_nonce or something like that ?

File: libraries/node/consensus/include/pbft/block_proposer.hpp [hunk 22]
```diff
@@ -70,10 +68,10 @@ class SortitionPropose : public ProposeModelFace {
...
-                std::shared_ptr<DagManager> dag_mgr, std::shared_ptr<TransactionManager> trx_mgr,
-                std::shared_ptr<DagBlockManager> dag_blk_mgr, std::shared_ptr<FinalChain> final_chain, addr_t node_addr,
-                secret_t node_sk, vrf_sk_t vrf_sk, logger::Logger log_time)
+  BlockProposer(BlockProposerConfig const& bp_config, VdfConfig const& vdf_config, std::shared_ptr<DagManager> dag_mgr,
+                std::shared_ptr<TransactionManager> trx_mgr, std::shared_ptr<DagBlockManager> dag_blk_mgr,
+                std::shared_ptr<FinalChain> final_chain, addr_t node_addr, secret_t node_sk, vrf_sk_t vrf_sk,
+                logger::Logger log_time)
       : bp_config_(bp_config),
         dag_mgr_(dag_mgr),
         trx_mgr_(trx_mgr),
```
Comment (JakubFornadel) [inline libraries/node/consensus/include/pbft/block_proposer.hpp:77]: all shared_ptrs should be moved here

File: libraries/node/consensus/src/final_chain/state_api_data.cpp [hunk 23]
```diff
@@ -0,0 +1,103 @@
+#include "final_chain/state_api_data.hpp"
+
+#include <libdevcore/CommonJS.h>
+
+#include "common/constants.hpp"
+
+namespace taraxa::state_api {
+
+TaraxaEVMError::TaraxaEVMError(string type, string msg) : runtime_error(move(msg)), type(move(type)) {}
```
Comment (JakubFornadel) [inline libraries/node/consensus/src/final_chain/state_api_data.cpp:9]: std::move, std::string

File: libraries/node/include/node/node.hpp [hunk 24]
```diff
@@ -69,7 +68,6 @@ class FullNode : public std::enable_shared_from_this<FullNode> {
   std::shared_ptr<DagBlockManager> dag_blk_mgr_;
   std::shared_ptr<TransactionManager> trx_mgr_;
   std::shared_ptr<Network> network_;
-  std::shared_ptr<TransactionOrderManager> trx_order_mgr_;
```
Comment (JakubFornadel) [inline libraries/node/include/node/node.hpp:72]: TransactionOrderManager is now used only in single unit test -> DagBlockTest.overlap. I would move this class declaration directly into the test or move files somewhere next to the test so it is not part of folder with source code that is actually used in node...

File: libraries/storage/CMakeLists.txt (hunk 9 above)
Comment (JakubFornadel) [reply libraries/storage/CMakeLists.txt:17]: yeah I think I would go with moving it into the dag_block

File: libraries/types/dag_block/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,10 @@
+set(HEADERS include/dag/block.hpp)
+set(SOURCES src/block.cpp)
+
+add_library(dag_block ${SOURCES} ${HEADERS})
+target_include_directories(dag_block PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
+target_link_libraries(dag_block PUBLIC vdf) 
+target_link_libraries(dag_block PRIVATE 
+    common
+    logger
```
Comment (JakubFornadel) [inline libraries/types/dag_block/CMakeLists.txt:9]: why do wee need to link dag_block against logger  ?

File: libraries/types/pbft_block/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+    src/block.cpp
+    src/sync_block.cpp
+)
+
+add_library(pbft_block ${SOURCES} ${HEADERS})
+target_include_directories(pbft_block PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
+target_link_libraries(pbft_block PUBLIC dag_block transaction)
+target_link_libraries(pbft_block PRIVATE 
+    common
+    logger
```
Comment (JakubFornadel) [inline libraries/types/pbft_block/CMakeLists.txt:15]: same question about logger

File: libraries/types/proposal_period_levels_map/CMakeLists.txt (hunk 14 above)
Comment (JakubFornadel) [reply libraries/types/proposal_period_levels_map/CMakeLists.txt:1]: so why and how was this resolved when it is still here ?

File: tests/dag_block_test.cpp [hunk 27]
```diff
@@ -194,29 +193,6 @@ TEST_F(DagBlockTest, push_and_pop) {
   blk_qu.stop();
 }
 
-TEST_F(DagBlockTest, overlap) {
```
Comment (JakubFornadel) [inline tests/dag_block_test.cpp:197]: Why is this test removed ?

File: libraries/node/include/node/node.hpp (hunk 24 above)
Comment (kstdl) [reply libraries/node/include/node/node.hpp:72]: As you wrote in https://github.com/Taraxa-project/taraxa-node/pull/928#discussion_r718244810 TransactionOrderManager wasn’t used anywhere, but that test. So I removed it and that related test. I don't think we need to test unused functionality

File: tests/dag_block_test.cpp (hunk 27 above)
Comment (kstdl) [reply tests/dag_block_test.cpp:197]: As you wrote in #928 (comment) TransactionOrderManager wasn’t used anywhere, but that test. So I removed it and that related test. I don't think we need to test unused functionality

File: libraries/types/dag_block/CMakeLists.txt (hunk 25 above)
Comment (kstdl) [reply libraries/types/dag_block/CMakeLists.txt:9]: There was include, but it was unused. Fixed

File: libraries/types/proposal_period_levels_map/CMakeLists.txt (hunk 14 above)
Comment (kstdl) [reply libraries/types/proposal_period_levels_map/CMakeLists.txt:1]: Because there is another comment about this issue. I think there is no need to leave few unresolved conversations on 1 item

File: tests/network_test.cpp [hunk 28]
```diff
@@ -102,9 +102,9 @@ TEST_F(NetworkTest, transfer_lot_of_blocks) {
   }
 
   // add one valid as last
-  auto dag_genesis = nodes[0]->getConfig().chain.dag_genesis_block.getHash();
-  vdf_sortition::VdfConfig vdf_config(node_cfgs[0].chain.vdf);
-  vdf_sortition::VdfSortition vdf(vdf_config, nodes[0]->getVrfSecretKey(), getRlpBytes(1));
+  auto dag_genesis = nodes[1]->getConfig().chain.dag_genesis_block.getHash();
+  VdfConfig vdf_config(node_cfgs[1].chain.vdf);
+  vdf_sortition::VdfSortition vdf(vdf_config, nodes[1]->getVrfSecretKey(), getRlpBytes(1));
```
[the remark this answers is not in the document]
Comment (kstdl) [reply tests/network_test.cpp:107]: Probably also a problem during the rebase. Changed it back

File: libraries/aleth/CMakeLists.txt (hunk 6 above)
Comment (kstdl) [reply libraries/aleth/CMakeLists.txt:1]: removed it

File: libraries/node/consensus/src/transaction_manager/transaction_manager.cpp (hunk 18 above)
Comment (kstdl) [reply libraries/node/consensus/src/transaction_manager/transaction_manager.cpp:130]: According to configs there is no situations when `network_transaction_interval` should be zero. So it is kind of unreachable code

File: libraries/node/consensus/src/transaction_manager/transaction_manager.cpp (hunk 17 above)
Comment (kstdl) [reply libraries/node/consensus/src/transaction_manager/transaction_manager.cpp:126]: returned it back

File: libraries/node/consensus/include/final_chain/data.hpp (hunk 20 above)
Comment (kstdl) [reply libraries/node/consensus/include/final_chain/data.hpp:43]: Can't remember why they were moved to the header file. They are needed for Ethereum compatibility if I understand it correctly. 
I moved this methods to cpp file and implement it the way uncles_hash method implemented

File: tests/dag_block_test.cpp (hunk 27 above)
Comment (JakubFornadel) [reply tests/dag_block_test.cpp:197]: true

File: libraries/node/consensus/src/transaction_manager/transaction_manager.cpp (hunk 18 above)
Comment (MatusKysel) [reply libraries/node/consensus/src/transaction_manager/transaction_manager.cpp]: ok, I see

**Review by MatusKysel — APPROVED** [review]

**Review by JakubFornadel — APPROVED** [review]
